src/OpenTelemetry.Exporter.InMemory/InMemoryMetricExporter.cs

```diff
@@ -0,0 +1,47 @@
+// <copyright file="InMemoryMetricExporter.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System.Collections.Generic;
+
+using OpenTelemetry.Metrics;
+
+namespace OpenTelemetry.Exporter
+{
+    public class InMemoryMetricExporter : BaseExporter<Metric>
```

Comment:
Why do we need a new exporter class?

Reply:
We need an exporter that can perform the translation from `Metric` to `ExportedMetric`, and this new exporter does that.

The current `InMemoryExporter` is generic and the exported Type must match the Type in the Batch.
https://github.com/open-telemetry/opentelemetry-dotnet/blob/d3c0c6ed614b1496d47e58a184f462be71e22558/src/OpenTelemetry.Exporter.InMemory/InMemoryExporter.cs#L21-L31

Reply:
Sorry, what's the real problem?

Reply:
The real problem is that users cannot depend on a normally exported `Metric` to be representative of the point in time it was exported.
Because the MetricApi reuses `Metric` instances (by design), an exported `Metric` can continue to be updated.

The `Metric` class itself cannot be deep-copied, therefore we're introducing a new class which can be representative of the `Metric` and a new Exporter to perform the translation.

Reply:
Something I don't like about this implementation is that it's a one-off class for a specific use case.
We could attempt to future the Exporter library by defining an InMemoryTranslationExporter<Tinput, Toutput>.
It's _still_ a new class, but this would be a single new class to hopefully prevent dozens of one-off classes.

Reply:
Hi @reyang, circling back to this.

The problem I'm trying to address is that `InMemoryExporter<T>` receives and exports the same type, in this case `Metric`.
I want to introduce a new type `ExportedMetric` which is our copy of the `Metric`.

I don't see a way to accomplish this copy without introducing a new Exporter.

- Option 1: the simplest way to do this is a one-off exporter which performs this copy; `InMemoryMetricExporter`
- Option 2: to prevent this and future one-off exporters, I could solve this with an `InMemoryTranslationExporter<Tinput, Toutput>`.
  In this case, the caller must provide a `Func<Tinput, Toutput>`. 
  
In either option, the caller would be the extension method `AddInMemoryExporter` which I will contribute in a follow up PR along with test changes.

Reply:
Can InMemoryExporter be improved to cover the case, so you don't have to introduce a new class?

Reply:
Yes, I could introduce a means to provide an different behavior to the Export method.
I'm struggling with the best way to add this to the PublicApi and also explain customers would use it.

On this same thought, I'm investigating a different approach. I want to open a draft PR to get more feedback.